toolchain/base/value_store.h

```diff
@@ -170,63 +140,63 @@ class ValueStore
     auto index_to_id = [](auto pair) -> std::pair<IdT, ConstRefType> {
       auto [index, value] = pair;
-      return std::pair<IdT, ConstRefType>(IdT(index), value);
+      return std::pair<IdT, ConstRefType>(index, value);
     };
-    return llvm::map_range(llvm::enumerate(values_), index_to_id);
+    return llvm::map_range(llvm::enumerate(values()), index_to_id);
   }
 
  private:
-  // Poison the entire contents of the value store. This is used to detect cases
-  // where references to elements in a value store are used across calls that
-  // might modify the store.
-  auto PoisonAll() const -> void {
-#if CARBON_POISON_VALUE_STORES
-    if (!all_poisoned_) {
-      __asan_poison_memory_region(values_.data(),
-                                  values_.size() * sizeof(values_[0]));
-      all_poisoned_ = true;
-    }
-#endif
-  }
-  // Unpoison the entire contents of the value store.
-  auto UnpoisonAll() const -> void {
-#if CARBON_POISON_VALUE_STORES
-    __asan_unpoison_memory_region(values_.data(),
-                                  values_.size() * sizeof(values_[0]));
-    all_poisoned_ = false;
-#endif
-  }
-  // Poison a single element.
-  auto PoisonElement([[maybe_unused]] int element) const -> void {
-#if CARBON_POISON_VALUE_STORES
-    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
-#endif
-  }
-  // Unpoison a single element.
-  auto UnpoisonElement([[maybe_unused]] int element) const -> void {
-#if CARBON_POISON_VALUE_STORES
-    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
-    all_poisoned_ = false;
-#endif
-  }
+  friend class ValueStoreRange<IdT>;
+
+  using ChunkType = Internal::ValueStoreChunk<IdT, ValueType>;
+
+  // Number of elements added to the store. The number should never exceed what
+  // fits in an `int32_t`, which is checked in non-optimized builds in
+  // IdToChunkIndices().
+  int32_t size_ = 0;
+
+  // Storage for the `ValueType` objects, indexed by the id. We use a vector of
+  // chunks of `ValueType` instead of just a vector of `ValueType` so that
+  // addresses of `ValueType` objects are stable. This allows the rest of the
+  // toolchain to hold references into `ValueStore` without having to worry
+  // about invalidation and use-after-free. We ensure at least one Chunk is held
+  // inline so that in the case where there is only a single Chunk (i.e. small
+  // files) we can avoid one indirection.
+  llvm::SmallVector<ChunkType, 1> chunks_;
+};
 
-  // Set inline size to 0 because these will typically be too large for the
-  // stack, while this does make File smaller.
-  llvm::SmallVector<std::decay_t<ValueType>, 0> values_;
+// A range over references to the values in a ValueStore, returned from
+// `ValueStore::values()`. Hides the complex type name of the iterator
+// internally to provide a type name (`ValueStoreRange<IdT>`) that can be
+// referred to without auto and templates.
```

Comment:
Out of curiosity, had you considered doing this directly on `ValueStore`, e.g.:

```
  auto values() -> auto { ... }
  using Range = decltype(MakeFlattenedRange(std::declval<const ValueStore<IdT>&>()));
```

And then instead of `ValueStoreRange<IdT>`, it's instead `ValueStore<IdT>::Range`

Reply:
I hadn't but I think I prefer code using `ValueStoreRange` as a function argument when it's required over a tested type in this case, since it's a full class type. Kind of like ArrayRef for SmallVector. It's probably just a matter of taste.

Not using `auto` is most of the point of `ValueStoreRange` so that it's possible to write a function receiving a range of values without making the function a template (like with ArrayRef before).